### PR TITLE
Core/Entities: add DB-support for conversation actorGuids

### PR DESCRIPTION
**Pull request by funjoker**
**Changes proposed:**

-  Added the option to add permanent spawns
-  Use only for "creature_text" like conversations.

**Target branch(es):** 3.3.5/master

- [ ] 3.3.5
- [x] master

**Issues addressed:** 

**Tests performed:** (Does it build, tested in-game, etc.)
Tested Malfurion Ysera conversation in Val'sharah, with sniffed conversation data.

wowscrnshot_103117_041118
wowscrnshot_103117_041125
wowscrnshot_103117_041132
wowscrnshot_103117_041143

**Known issues and TODO list:** (add/remove lines as needed)

- [ ] Whatever TC wants
- [ ]

### CODE REVIEW

File: src/server/game/Entities/Conversation/Conversation.cpp [hunk 1]
```diff
@@ -130,7 +131,22 @@ bool Conversation::Create(ObjectGuid::LowType lowGuid, uint32 conversationEntry,
             actorField.Type = ConversationDynamicFieldActor::ActorType::CreatureActor;
             SetDynamicStructuredValue(CONVERSATION_DYNAMIC_FIELD_ACTORS, actorsIndex++, &actorField);
         }
-        else
+        else if (!conversationTemplate->Actors.empty())
```
Comment (joschiwald) [inline src/server/game/Entities/Conversation/Conversation.cpp]: this won't work if you have a mix of both actor types, you need to start from 0 when adding actors with guid

File: src/server/game/Globals/ConversationDataStore.cpp [hunk 2]
```diff
@@ -103,18 +105,37 @@ void ConversationDataStore::LoadConversationTemplates()
...
             {
+                if (actorGuid != 0)
+                    TC_LOG_ERROR("sql.sql", "Table `conversation_actors` references both actor (ID: %u) and actorGuid (GUID: " UI64FMTD ") for Conversation %u, GUID is ignored", actorId, actorGuid, conversationId);
+
                 std::vector<ConversationActorTemplate const*>& actors = actorsByConversation[conversationId];
                 if (actors.size() <= idx)
                     actors.resize(idx + 1);
                 actors[idx] = conversationActorTemplate;
-                ++count;
+                ++count;
```
Comment (joschiwald) [inline src/server/game/Globals/ConversationDataStore.cpp]: whitespace

File: src/server/game/Globals/ConversationDataStore.cpp [hunk 3]
```diff
@@ -103,18 +105,37 @@ void ConversationDataStore::LoadConversationTemplates()
...
             uint32 conversationId = fields[0].GetUInt32();
             uint32 actorId = fields[1].GetUInt32();
-            uint16 idx = fields[2].GetUInt16();
+            uint64 actorGuid = fields[2].GetUInt64();
+            uint16 idx = fields[3].GetUInt16();
 
             if (ConversationActorTemplate const* conversationActorTemplate = Trinity::Containers::MapGetValuePtr(_conversationActorTemplateStore, actorId))
             {
+                if (actorGuid != 0)
+                    TC_LOG_ERROR("sql.sql", "Table `conversation_actors` references both actor (ID: %u) and actorGuid (GUID: " UI64FMTD ") for Conversation %u, GUID is ignored", actorId, actorGuid, conversationId);
```
Comment (joschiwald) [inline src/server/game/Globals/ConversationDataStore.cpp]: move this message above the if condition, and check what your message says

File: src/server/game/Globals/ConversationDataStore.cpp [hunk 4]
```diff
@@ -103,18 +105,37 @@ void ConversationDataStore::LoadConversationTemplates()
...
+
                 std::vector<ConversationActorTemplate const*>& actors = actorsByConversation[conversationId];
                 if (actors.size() <= idx)
                     actors.resize(idx + 1);
                 actors[idx] = conversationActorTemplate;
-                ++count;
+                ++count;                
             }
             else
                 TC_LOG_ERROR("sql.sql", "Table `conversation_actors` references an invalid actor (ID: %u) for Conversation %u, skipped", actorId, conversationId);
```
Comment (joschiwald) [inline src/server/game/Globals/ConversationDataStore.cpp]: you will print this error for all actors with guid

File: src/server/game/Entities/Conversation/Conversation.cpp [hunk 5]
```diff
@@ -19,6 +19,7 @@
 #include "Map.h"
 #include "Unit.h"
 #include "UpdateData.h"
+#include "Creature.h"
```
Comment (Vincent-Michael) [inline src/server/game/Entities/Conversation/Conversation.cpp]: really need?

File: src/server/game/Globals/ConversationDataStore.cpp [hunk 6]
```diff
@@ -103,18 +105,37 @@ void ConversationDataStore::LoadConversationTemplates()
 
             uint32 conversationId = fields[0].GetUInt32();
             uint32 actorId = fields[1].GetUInt32();
-            uint16 idx = fields[2].GetUInt16();
+            uint64 actorGuid = fields[2].GetUInt64();
```
Comment (joschiwald) [inline src/server/game/Globals/ConversationDataStore.cpp]: ObjectGuid::LowType

File: src/server/game/Globals/ConversationDataStore.cpp (hunk 3 above)
Comment (funjoker) [reply src/server/game/Globals/ConversationDataStore.cpp]: It shall just print this error.
Due to other if clause the guid is ignored, See some lines below.

File: src/server/game/Entities/Conversation/Conversation.cpp (hunk 5 above)
Comment (funjoker) [reply src/server/game/Entities/Conversation/Conversation.cpp]: bounds.first->second->GetGUID()

File: src/server/game/Globals/ConversationDataStore.cpp (hunk 6 above)
Comment (funjoker) [reply src/server/game/Globals/ConversationDataStore.cpp]: Ok

File: src/server/game/Globals/ConversationDataStore.cpp [hunk 7]
```diff
@@ -103,18 +105,37 @@ void ConversationDataStore::LoadConversationTemplates()
...
+                if (creData)
+                {
+                    std::vector<ObjectGuid::LowType>& guids = actorGuidsByConversation[conversationId];
+                    if (guids.size() <= idx)
+                        guids.resize(idx + 1);
+                    guids[idx] = actorGuid;
+                    ++count;
+                }
+                else
+                    TC_LOG_ERROR("sql.sql","Table `conversation_actors` references an invalid creature guid (GUID: " UI64FMTD ") for Conversation %u, skipped", actorGuid, conversationId);
```
Comment (joschiwald) [inline src/server/game/Globals/ConversationDataStore.cpp]: missing space after comma

File: src/server/game/Entities/Conversation/Conversation.cpp (hunk 1 above)
Comment (funjoker) [reply src/server/game/Entities/Conversation/Conversation.cpp]: you are right. Gonna fix that

File: src/server/game/Globals/ConversationDataStore.cpp (hunk 4 above)
Comment (funjoker) [reply src/server/game/Globals/ConversationDataStore.cpp]: will adjust all Logs when at home thank you

File: src/server/game/Entities/Conversation/Conversation.cpp (hunk 5 above)
Comment (Shauren) [reply src/server/game/Entities/Conversation/Conversation.cpp]: sort includes alphabetically, keeping `Conversation.h` first

File: src/server/game/Entities/Conversation/Conversation.cpp [hunk 8]
```diff
@@ -130,7 +131,22 @@ bool Conversation::Create(ObjectGuid::LowType lowGuid, uint32 conversationEntry,
...
+            ++actorsIndex;
+    }
+
+    for (ObjectGuid::LowType actorGuids : conversationTemplate->ActorGuids)
+    {
+        if (actorGuids)
+        {
+            auto bounds = map->GetCreatureBySpawnIdStore().equal_range(actorGuids);
+
+            if (bounds.first == bounds.second)
```
Comment (Shauren) [inline src/server/game/Entities/Conversation/Conversation.cpp]: This isnt (yet) part of master branch but having dynamic spawning in mind, there could be multiple spawns for the same db guid and you probably should use the alive/last one

File: src/server/game/Globals/ConversationDataStore.cpp [hunk 9]
```diff
@@ -21,6 +21,7 @@
 #include "DB2Stores.h"
 #include "Log.h"
 #include "Timer.h"
+#include "ObjectMgr.h"
```
Comment (Shauren) [inline src/server/game/Globals/ConversationDataStore.cpp]: sort

**Review by Shauren — CHANGES_REQUESTED** [review]

File: src/server/game/Entities/Conversation/Conversation.cpp [hunk 10]
```diff
@@ -134,6 +136,28 @@ bool Conversation::Create(ObjectGuid::LowType lowGuid, uint32 conversationEntry,
...
 
+    // reset actorsIndex
+    actorsIndex = 0;
+
+    for (ObjectGuid::LowType actorGuids : conversationTemplate->ActorGuids)
+    {
+        if (actorGuids)
+        {
+            auto bounds = map->GetCreatureBySpawnIdStore().equal_range(actorGuids);
+
```
Comment (funjoker) [inline src/server/game/Entities/Conversation/Conversation.cpp]: What is the best way to get the creature, checking it's alive, compatible with the dynamic_spawn system ?

**Comment by funjoker** [comment]
#20758 
contains checks if actors are fully set.

Both together will make it fit.

File: sql/updates/world/master/9999_99_99_99_world.sql [hunk 11]
```diff
@@ -0,0 +1,5 @@
+ALTER TABLE `conversation_actors`
+	CHANGE COLUMN `ConversationActorId` `ConversationActorId` INT(10) UNSIGNED NOT NULL DEFAULT '0' AFTER `ConversationId`,
+	ADD COLUMN `ConversationActorGuid` BIGINT(20) UNSIGNED NOT NULL DEFAULT '0' AFTER `ConversationActorId`,
+	DROP PRIMARY KEY,
+	ADD PRIMARY KEY (`ConversationId`, `ConversationActorId`, `ConversationActorGuid`);
```
Comment (joschiwald) [inline sql/updates/world/master/9999_99_99_99_world.sql]: what about pk with `ConversationId` and `Idx`?

**Pull request merged**